Best-effort Error Detection for Using Deleted UserRRefs

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34673 Best-effort Error Detection for Using Deleted UserRRefs**
* #34681 Use c10::str in py_rref.cpp
* #34679 Use c10::str in process_group_agent.cpp

Differential Revision: [D20427839](https://our.internmc.facebook.com/intern/diff/D20427839)

mrshenli committed:
Best-effort Error Detection for Using Deleted UserRRefs

[ghstack-poisoned]

mrshenli committed:
Update on "Best-effort Error Detection for Using Deleted UserRRefs"




[ghstack-poisoned]

mrshenli committed:
Update on "Best-effort Error Detection for Using Deleted UserRRefs"


Differential Revision: [D20427839](https://our.internmc.facebook.com/intern/diff/D20427839)

[ghstack-poisoned]

mrshenli committed:
Update on "Best-effort Error Detection for Using Deleted UserRRefs"


Differential Revision: [D20427839](https://our.internmc.facebook.com/intern/diff/D20427839)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -143,6 +152,38 @@ IValue UserRRef::toHere() {
+      forkId(),
+      " has been deleted. Cannot call fork an UserRRef after deletion.");
+  return RRef::fork();
+}
```
Not asking for changes, but what could be a robust way to avoid forking on graceful shutdown. A similar thing to agent.sync() to flush all tasks in the queues on all workers?

mrshenli replied:
Yep, it would at least require a global flushing, which we are trying to get rid of. Another thing is we don't have anything to prevent applications to create a local thread and use RRef/RPC in that thread. So it possible that the main thread is calling shutdown, but another thread is fork an RRef concurrently?

xush6528 commented on torch/csrc/distributed/rpc/rref_impl.cpp:
```diff
@@ -105,6 +105,15 @@ const ForkId& UserRRef::forkId() const {
+      rrefId(),
+      " and ForkId=",
+      forkId(),
+      " has been deleted. Cannot call to_here() on it after deletion.");
```
This should be a robust check? since after entering shutdown(gracefuk) and run past _wait_all_workers(), there should be no to_here() calls.

more to_here() calls could happen after shutdown(), but the deletedOnOwner_ status should be correct.

mrshenli replied:
I agree, if we assume application code is single-threaded.

mrshenli committed:
Update on "Best-effort Error Detection for Using Deleted UserRRefs"


Differential Revision: [D20427839](https://our.internmc.facebook.com/intern/diff/D20427839)

[ghstack-poisoned]
